File: autocapture/dewarp/homography.py
"""Pure-numpy 3x3 homography + bilinear inverse-warp.

This module has no dependencies beyond numpy. It is the verified fallback
for when OpenCV is not available. In production, prefer
``cv2.getPerspectiveTransform`` + ``cv2.warpPerspective`` for speed.

Homography math: a 3x3 matrix H maps a point (x, y) in the source image
to (x', y') in the destination via

    [x' y' w']^T  =  H @ [x y 1]^T
    output        =  (x'/w', y'/w')

We solve for H from 4 point correspondences using the Direct Linear
Transform (DLT) method. Then we apply H with inverse mapping + bilinear
interpolation so the output is fully sampled and has no holes.
"""

from __future__ import annotations
import numpy as np


def compute_homography(src_pts: np.ndarray, dst_pts: np.ndarray) -> np.ndarray:
    """Compute 3x3 homography mapping src_pts -> dst_pts.

    Each input is shape (4, 2). We solve for H by enforcing the
    forward map exactly on the 4 correspondences and on the centroid,
    then disambiguating the projective sign by trying both
    orientations of the linear part and keeping the one whose
    determinant is positive (forward map, not reflection).

    Returns
    -------
    H : (3, 3) ndarray
    """
    assert src_pts.shape == (4, 2), f"src_pts must be (4,2), got {src_pts.shape}"
    assert dst_pts.shape == (4, 2), f"dst_pts must be (4,2), got {dst_pts.shape}"

    # Direct Linear Transform: 8 equations in 8 unknowns.
    A = np.zeros((8, 9), dtype=np.float64)
    for i, ((sx, sy), (dx, dy)) in enumerate(zip(src_pts, dst_pts)):
        A[2 * i    ] = [0, 0, 0, -sx, -sy, -1,  dy * sx,  dy * sy,  dy]
        A[2 * i + 1] = [-sx, -sy, -1, 0, 0, 0, -dx * sx, -dx * sy, -dx]

    # SVD: H is the right-singular vector for the smallest singular value.
    _, _, Vt = np.linalg.svd(A)
    H = Vt[-1].reshape(3, 3)

    # Normalize so H[2, 2] = 1
    if abs(H[2, 2]) > 1e-12:
        H = H / H[2, 2]

    # Sign + reflection disambiguation. The DLT has 4 solutions that
    # are projectively equivalent (differing by a sign or by
    # reflection across an axis). We test all 4 candidates: H, -H, and
    # the two single-axis reflections (negate H[0, :] or H[1, :]).
    diag_x = np.diag([-1.0, 1.0, 1.0])
    diag_y = np.diag([1.0, -1.0, 1.0])
    candidates = [H, -H, diag_x @ H, diag_y @ H]

    # Test on the 4 correspondences (zero error in all) and the centroid
    # (positive mass if forward).
    src_c = src_pts.mean(axis=0)
    dst_c = dst_pts.mean(axis=0)

    best = None
    best_score = -np.inf
    for c in candidates:
        proj = c @ np.array([src_c[0], src_c[1], 1.0])
        cx, cy = proj[0] / proj[2], proj[1] / proj[2]
        # Score: forward-map bonus (positive 2x2 det) + low centroid error
        det2 = np.linalg.det(c[:2, :2])
        score = det2 * 1e6 - ((cx - dst_c[0]) ** 2 + (cy - dst_c[1]) ** 2)
        if score > best_score:
            best_score = score
            best = c
    return best


def warp_bilinear(
    src: np.ndarray,
    H: np.ndarray,
    out_h: int,
    out_w: int,
) -> np.ndarray:
    """Inverse-warp src through H to produce an (out_h, out_w) image.

    For every output pixel (u, v), we solve for the corresponding
    source pixel (x, y) by applying H^-1, then sample src with bilinear
    interpolation. Pixels that fall outside src are filled with zeros.

    Parameters
    ----------
    src : (H, W, C) or (H, W) ndarray
    H   : (3, 3) homography, src -> dst
    out_h, out_w : output dimensions

    Returns
    -------
    dst : (out_h, out_w, C) or (out_h, out_w) ndarray
    """
    assert src.ndim in (2, 3), f"src must be 2D or 3D, got {src.ndim}D"
    H_inv = np.linalg.inv(H)

    H_src, W_src = src.shape[:2]
    channels = src.shape[2] if src.ndim == 3 else 1
    if src.ndim == 2:
        src = src[..., None]

    # Build the destination coordinate grid
    u, v = np.meshgrid(np.arange(out_w, dtype=np.float64),
                       np.arange(out_h, dtype=np.float64))

    # Apply H^-1 to every (u, v, 1)
    ones = np.ones_like(u)
    homog = np.stack([u, v, ones], axis=-1)        # (out_h, out_w, 3)
    src_h = homog @ H_inv.T                        # (out_h, out_w, 3)
    src_w = src_h[..., 2]
    # Guard against exact-zero denominators
    src_w = np.where(np.abs(src_w) < 1e-12, 1e-12, src_w)
    x = src_h[..., 0] / src_w
    y = src_h[..., 1] / src_w

    # Bilinear sample
    x0 = np.floor(x).astype(np.int64)
    y0 = np.floor(y).astype(np.int64)
    x1 = x0 + 1
    y1 = y0 + 1

    wx = x - x0
    wy = y - y0

    # Validity: the floor coordinates must be in-range. x1/y1 are
    # allowed to be one past the edge (we clip them and use the
    # edge value via wx/wy = 0). This is what makes an integer
    # coordinate like x=19 sample exactly src[19] instead of 0.
    valid = (x0 >= 0) & (x0 < W_src) & (y0 >= 0) & (y0 < H_src)
    x0c = np.clip(x0, 0, W_src - 1)
    x1c = np.clip(x1, 0, W_src - 1)
    y0c = np.clip(y0, 0, H_src - 1)
    y1c = np.clip(y1, 0, H_src - 1)

    Ia = src[y0c, x0c]   # top-left
    Ib = src[y0c, x1c]   # top-right
    Ic = src[y1c, x0c]   # bottom-left
    Id = src[y1c, x1c]   # bottom-right

    wx_e = wx[..., None] if src.ndim == 3 else wx
    wy_e = wy[..., None] if src.ndim == 3 else wy

    top = Ia * (1 - wx_e) + Ib * wx_e
    bot = Ic * (1 - wx_e) + Id * wx_e
    out = top * (1 - wy_e) + bot * wy_e

    out = np.where(valid[..., None] if src.ndim == 3 else valid,
                   out,
                   0.0)
    if src.ndim == 2:
        out = out[..., 0]
    return out.astype(src.dtype)


def warp_quad(src: np.ndarray, quad: np.ndarray, target_aspect: str = "A4") -> np.ndarray:
    """Warp src so that the detected 4-corner quad becomes a rectangle.

    Parameters
    ----------
    quad : (4, 2) ndarray, ordered TL, TR, BR, BL
    target_aspect : "A4" (sqrt(2)), "US_LETTER" (11/8.5), "MAX_EDGE" (square),
                    or "AUTO" (average of quad edges)
    """
    pts = quad.astype(np.float64)
    # edge lengths
    top    = np.linalg.norm(pts[1] - pts[0])
    bot    = np.linalg.norm(pts[2] - pts[3])
    right  = np.linalg.norm(pts[2] - pts[1])
    left   = np.linalg.norm(pts[0] - pts[3])
    width  = max(top, bot)
    height = max(left, right)

    if target_aspect == "A4":
        aspect = np.sqrt(2.0)
        if height > width / aspect:
            height = int(width / aspect)
        else:
            width = int(height * aspect)
    elif target_aspect == "US_LETTER":
        aspect = 11.0 / 8.5
        if height > width / aspect:
            height = int(width / aspect)
        else:
            width = int(height * aspect)
    elif target_aspect == "MAX_EDGE":
        side = int(max(width, height))
        width, height = side, side
    else:  # AUTO
        pass

    width  = int(width)
    height = int(height)
    dst = np.array([[0, 0], [width - 1, 0],
                    [width - 1, height - 1], [0, height - 1]],
                   dtype=np.float64)
    H = compute_homography(pts, dst)
    return warp_bilinear(src, H, height, width)

File: src/dewarp/homography.js
// 3x3 homography + bilinear inverse-warp.
//
// Direct port of autocapture/dewarp/homography.py.
//
// Math: a homography H maps (x,y) -> (x',y') in projective coords.
// Solving for H from 4 correspondences uses the Direct Linear Transform
// (DLT) — 8 equations in 8 unknowns, solved by SVD on an 8x9 matrix.
// The right-singular vector with the smallest singular value is H.

import { svdRightVectors } from "../math/svd.js";
import { det2x2 } from "../math/ndarray.js";

/**
 * Compute the 3x3 homography mapping src_pts (4,2) -> dst_pts (4,2).
 * @returns {Float64Array} length-9 row-major matrix
 */
export function computeHomography(srcPts, dstPts) {
  if (srcPts.length !== 8 || dstPts.length !== 8) {
    throw new Error(`computeHomography: expected flat length-8 arrays, got ${srcPts.length} / ${dstPts.length}`);
  }
  const A = new Float64Array(8 * 9);
  for (let i = 0; i < 4; i++) {
    const sx = srcPts[i * 2], sy = srcPts[i * 2 + 1];
    const dx = dstPts[i * 2], dy = dstPts[i * 2 + 1];
    // row 2i
    A[(2 * i) * 9 + 0] = 0;      A[(2 * i) * 9 + 1] = 0;      A[(2 * i) * 9 + 2] = 0;
    A[(2 * i) * 9 + 3] = -sx;    A[(2 * i) * 9 + 4] = -sy;    A[(2 * i) * 9 + 5] = -1;
    A[(2 * i) * 9 + 6] = dy * sx; A[(2 * i) * 9 + 7] = dy * sy; A[(2 * i) * 9 + 8] = dy;
    // row 2i+1
    A[(2 * i + 1) * 9 + 0] = -sx; A[(2 * i + 1) * 9 + 1] = -sy; A[(2 * i + 1) * 9 + 2] = -1;
    A[(2 * i + 1) * 9 + 3] = 0;   A[(2 * i + 1) * 9 + 4] = 0;   A[(2 * i + 1) * 9 + 5] = 0;
    A[(2 * i + 1) * 9 + 6] = -dx * sx; A[(2 * i + 1) * 9 + 7] = -dx * sy; A[(2 * i + 1) * 9 + 8] = -dx;
  }

  // SVD: smallest singular value's right-singular vector is the last row of Vt
  const Vt = svdRightVectors(A, 8, 9);
  // Our svdRightVectors sorts rows of V ascending by sigma, so Vt[0]
  // is the right-singular vector for the smallest singular value
  // (matches the DLT requirement). np.linalg.svd sorts descending,
  // which is why the python implementation uses Vt[-1] — same row.
  const H = new Float64Array(9);
  for (let i = 0; i < 9; i++) H[i] = Vt[0 * 9 + i];  // first row

  // Normalize so H[2,2] = 1
  if (Math.abs(H[8]) > 1e-12) {
    for (let i = 0; i < 9; i++) H[i] /= H[8];
  }

  // Disambiguate the 4 projective solutions: try H, -H, and the two
  // single-axis row reflections. The DLT gives H up to scale and up to
  // a 4-element Klein-4 ambiguity {I, -I, diag(-1,1,1), diag(1,-1,1)}
  // acting on the LEFT (negates a row). Pick the one whose 2x2 linear
  // part has positive determinant and minimal centroid error.
  const candidates = [
    H,
    H.map((v) => -v),
    reflectRow(H, 0, -1),  // diag(-1, 1, 1) @ H: negate row 0
    reflectRow(H, 1, -1),  // diag(1, -1, 1) @ H: negate row 1
  ];

  const srcCx = (srcPts[0] + srcPts[2] + srcPts[4] + srcPts[6]) / 4;
  const srcCy = (srcPts[1] + srcPts[3] + srcPts[5] + srcPts[7]) / 4;
  const dstCx = (dstPts[0] + dstPts[2] + dstPts[4] + dstPts[6]) / 4;
  const dstCy = (dstPts[1] + dstPts[3] + dstPts[5] + dstPts[7]) / 4;

  let best = null, bestScore = -Infinity;
  for (const c of candidates) {
    // c @ [srcCx, srcCy, 1]
    const px = c[0] * srcCx + c[1] * srcCy + c[2];
    const py = c[3] * srcCx + c[4] * srcCy + c[5];
    const pw = c[6] * srcCx + c[7] * srcCy + c[8];
    if (Math.abs(pw) < 1e-12) continue;
    const cx = px / pw;
    const cy = py / pw;
    // det of top-left 2x2
    const detTopLeft = det2x2([c[0], c[1], c[3], c[4]]);
    const score = detTopLeft * 1e6 - ((cx - dstCx) ** 2 + (cy - dstCy) ** 2);
    if (score > bestScore) { bestScore = score; best = c; }
  }
  return best;
}

/** Multiply on the left by diag(...,1,...) but with sign flip on row k. */
function reflectRow(H, k, sign) {
  const out = new Float64Array(H);
  for (let j = 0; j < 3; j++) out[k * 3 + j] *= sign;
  return out;
}

/**
 * Inverse-warp src through H to produce an (outH, outW) image.
 *
 * For each output pixel (u, v) we apply H^-1 to get the source (x, y),
 * then sample with bilinear interpolation. Pixels outside src fill with 0.
 *
 * @param {Uint8ClampedArray|Uint8Array} src length H_src*W_src*C
 * @param {number} H_src
 * @param {number} W_src
 * @param {number} C channels (1 or 3)
 * @param {Float64Array} H length-9, src->dst
 * @param {number} outH
 * @param {number} outW
 * @returns {Uint8ClampedArray} length outH*outW*C
 */
export function warpBilinear(src, H_src, W_src, C, H, outH, outW) {
  // Inverse H
  const Hi = invert3x3(H);

  const out = new Uint8ClampedArray(outH * outW * C);
  const srcH = H_src, srcW = W_src;

  for (let v = 0; v < outH; v++) {
    for (let u = 0; u < outW; u++) {
      // Hi @ [u, v, 1]
      const hx = Hi[0] * u + Hi[1] * v + Hi[2];
      const hy = Hi[3] * u + Hi[4] * v + Hi[5];
      const hw = Hi[6] * u + Hi[7] * v + Hi[8];
      const denom = Math.abs(hw) < 1e-12 ? 1e-12 : hw;
      const x = hx / denom;
      const y = hy / denom;

      const x0 = Math.floor(x);
      const y0 = Math.floor(y);
      if (x0 < 0 || x0 >= srcW || y0 < 0 || y0 >= srcH) continue;  // leaves 0
      const x1 = Math.min(srcW - 1, x0 + 1);
      const y1 = Math.min(srcH - 1, y0 + 1);
      const wx = x - x0;
      const wy = y - y0;
      const base = (v * outW + u) * C;
      for (let c = 0; c < C; c++) {
        const ia = src[(y0 * srcW + x0) * C + c];
        const ib = src[(y0 * srcW + x1) * C + c];
        const ic = src[(y1 * srcW + x0) * C + c];
        const id = src[(y1 * srcW + x1) * C + c];
        const top = ia * (1 - wx) + ib * wx;
        const bot = ic * (1 - wx) + id * wx;
        out[base + c] = Math.round(top * (1 - wy) + bot * wy);
      }
    }
  }
  return out;
}

function invert3x3(M) {
  const a = M[0], b = M[1], c = M[2];
  const d = M[3], e = M[4], f = M[5];
  const g = M[6], h = M[7], i = M[8];
  const A = e * i - f * h;
  const B = -(d * i - f * g);
  const C = d * h - e * g;
  const D = -(b * i - c * h);
  const E = a * i - c * g;
  const F = -(a * h - b * g);
  const G = b * f - c * e;
  const Hc = -(a * f - c * d);
  const Ic = a * e - b * d;
  const det = a * A + b * B + c * C;
  const invDet = 1.0 / det;
  return new Float64Array([
    A * invDet, D * invDet, G * invDet,
    B * invDet, E * invDet, Hc * invDet,
    C * invDet, F * invDet, Ic * invDet,
  ]);
}

/**
 * Warp src so the 4-corner quad becomes a rectangle.
 *
 * @param {Uint8ClampedArray} src
 * @param {number} srcH
 * @param {number} srcW
 * @param {number} C
 * @param {number[]} quad flat 8-array, TL TR BR BL
 * @param {string} targetAspect "A4" | "US_LETTER" | "MAX_EDGE" | "AUTO"
 * @returns {{ data: Uint8ClampedArray, width: number, height: number }}
 */
export function warpQuad(src, srcH, srcW, C, quad, targetAspect = "A4") {
  // Edge lengths
  const top    = dist(quad[0], quad[1], quad[2], quad[3]);
  const bot    = dist(quad[6], quad[7], quad[4], quad[5]);
  const right  = dist(quad[2], quad[3], quad[4], quad[5]);
  const left   = dist(quad[0], quad[1], quad[6], quad[7]);
  let width = Math.max(top, bot);
  let height = Math.max(left, right);

  if (targetAspect === "A4") {
    const aspect = Math.sqrt(2.0);
    if (height > width / aspect) height = Math.floor(width / aspect);
    else width = Math.floor(height * aspect);
  } else if (targetAspect === "US_LETTER") {
    const aspect = 11.0 / 8.5;
    if (height > width / aspect) height = Math.floor(width / aspect);
    else width = Math.floor(height * aspect);
  } else if (targetAspect === "MAX_EDGE") {
    const side = Math.floor(Math.max(width, height));
    width = side; height = side;
  }
  width = Math.max(1, Math.floor(width));
  height = Math.max(1, Math.floor(height));

  const dst = [0, 0, width - 1, 0, width - 1, height - 1, 0, height - 1];
  const H = computeHomography(quad, dst);
  const data = warpBilinear(src, srcH, srcW, C, H, height, width);
  return { data, width, height };
}

function dist(x1, y1, x2, y2) {
  const dx = x1 - x2, dy = y1 - y2;
  return Math.sqrt(dx * dx + dy * dy);
}